### PR TITLE
Add hook to set custom span name in flask instrumentation

### PR DESCRIPTION
**Pull request by lonewolf3739**
Addresses #134

### CODE REVIEW

File: instrumentation/opentelemetry-instrumentation-flask/src/opentelemetry/instrumentation/flask/__init__.py [hunk 1]
```diff
@@ -75,6 +75,15 @@ def get_excluded_urls():
 _excluded_urls = get_excluded_urls()
 
 
+def get_default_span_name():
+    span_name = None
```
Comment (lzchen) [inline instrumentation/opentelemetry-instrumentation-flask/src/opentelemetry/instrumentation/flask/__init__.py:79]: Nit: span_name = ""

**Comment by lzchen** [comment]
CHANGELOG entry?

**Review by lzchen — APPROVED** [review]
Nice! Just a couple of comments.

**Comment by lonewolf3739** [comment]
> CHANGELOG entry?

I missed that, added now.

File: instrumentation/opentelemetry-instrumentation-flask/src/opentelemetry/instrumentation/flask/__init__.py [hunk 2]
```diff
@@ -105,43 +114,40 @@ def _start_response(status, response_headers, *args, **kwargs):
...
-            attributes["http.route"] = flask.request.url_rule.rule
-        for key, value in attributes.items():
-            span.set_attribute(key, value)
-
-    activation = tracer.use_span(span, end_on_exit=True)
-    activation.__enter__()
-    environ[_ENVIRON_ACTIVATION_KEY] = activation
-    environ[_ENVIRON_SPAN_KEY] = span
-    environ[_ENVIRON_TOKEN] = token
+        environ = flask.request.environ
```
Comment (lzchen) [inline instrumentation/opentelemetry-instrumentation-flask/src/opentelemetry/instrumentation/flask/__init__.py:122]: I think we should pass in `environ` as an argument to the span callback, so users can change the span name based off these parameters.

**Review by lzchen — CHANGES_REQUESTED** [review]
Just a comment.

File: instrumentation/opentelemetry-instrumentation-flask/src/opentelemetry/instrumentation/flask/__init__.py (hunk 2 above)
Comment (lonewolf3739) [reply instrumentation/opentelemetry-instrumentation-flask/src/opentelemetry/instrumentation/flask/__init__.py:122]: Yes, I thought about it. The request context `flask.request` is accessible for users just like how we are using it in instrumentation. Passing `environ` wouldn't really give any extra benefit. What do you think?

Comment (lzchen) [reply instrumentation/opentelemetry-instrumentation-flask/src/opentelemetry/instrumentation/flask/__init__.py:122]: Ahh that is true. Yeah maybe we don't need it in the params then.

**Review by lzchen — APPROVED** [review]

File: instrumentation/opentelemetry-instrumentation-flask/src/opentelemetry/instrumentation/flask/__init__.py [hunk 3]
```diff
@@ -188,14 +196,18 @@ def _instrument(self, **kwargs):
         self._original_flask = flask.Flask
```
Comment (lzchen) [inline instrumentation/opentelemetry-instrumentation-flask/src/opentelemetry/instrumentation/flask/__init__.py:196]: What if user instruments with `instrument()`? We should provide this functionality for that too.

**Review by lzchen — CHANGES_REQUESTED** [review]
a comment

File: instrumentation/opentelemetry-instrumentation-flask/src/opentelemetry/instrumentation/flask/__init__.py (hunk 3 above)
Comment (lonewolf3739) [reply instrumentation/opentelemetry-instrumentation-flask/src/opentelemetry/instrumentation/flask/__init__.py:196]: The current implementation is using `get_default_span_name`. I am wondering how do we hook the name call back in this case?

Comment (lzchen) [reply instrumentation/opentelemetry-instrumentation-flask/src/opentelemetry/instrumentation/flask/__init__.py:196]: Similar to [requests](https://github.com/open-telemetry/opentelemetry-python-contrib/pull/158/files), use a kwarg.

Comment (lonewolf3739) [reply instrumentation/opentelemetry-instrumentation-flask/src/opentelemetry/instrumentation/flask/__init__.py:201]: I added it. One thing is we are patching the `flask.Flask` to get `instrument()` to work but if an import of Flask is made before the `FlaskInstrumentor().instrument()` then it doesn't do anything. I don't think if this is documented somewhere.

Comment (lzchen) [reply instrumentation/opentelemetry-instrumentation-flask/src/opentelemetry/instrumentation/flask/__init__.py:201]: Yes this is known behaviour.

**Review by lzchen — APPROVED** [review]

**Pull request merged**